streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamWindowAggregate.java

```diff
@@ -184,6 +247,75 @@ public void process(final Record<KIn, VIn> record) {
+            // Throttle emit frequency
+            if (now < timeTracker.nextTimeToEmit) {
+                return;
+            }
+
+            // Schedule next emit time based on now to avoid the case that if system time jumps a lot,
+            // this can be triggered everytime
+            timeTracker.nextTimeToEmit = now;
+            timeTracker.advanceNextTimeToEmit();
+
+            // Close time does not progress
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP && lastEmitCloseTime >= closeTime) {
+                return;
+            }
+
+            final long emitWindowStart = closeTime - windows.size();
+            if (emitWindowStart < 0) {
+                // If emitWindowStart is 0, it means first window closes since windowEndTime
+                // is exclusive
+                return;
+            }
+
+            // Because we only get here when emitWindowStart > 0 which means closeTime > windows.size()
+            // Since we set lastEmitCloseTime to closeTime before storing to processor metadata
+            // lastEmitCloseTime - windows.size() is always > 0
+            // Set lastEmitWindowStart to -1L if not set so that when we fetchAll, we fetch from 0L
+            final long lastEmitWindowStart = lastEmitCloseTime == ConsumerRecord.NO_TIMESTAMP ?
+                -1L : lastEmitCloseTime - windows.size();
+
+            if (lastEmitCloseTime != ConsumerRecord.NO_TIMESTAMP) {
+                final Map<Long, W> matchedCloseWindows = windows.windowsFor(emitWindowStart);
+                final Map<Long, W> matchedEmitWindows = windows.windowsFor(lastEmitWindowStart);
+
+                // Don't fetch store if the new emit window close time doesn't progress enough to cover next
+                // window
+                if (matchedCloseWindows.equals(matchedEmitWindows)) {
+                    log.debug("no new windows to emit. LastEmitCloseTime={}, newCloseTime={}",
+                        lastEmitCloseTime, closeTime);
+                    return;
+                }
+            }
+
+            final KeyValueIterator<Windowed<KIn>, ValueAndTimestamp<VAgg>> windowToEmit =  windowStore
+                .fetchAll(lastEmitWindowStart + 1, emitWindowStart);
+
+            int emittedCount = 0;
+            while (windowToEmit.hasNext()) {
+                emittedCount++;
+                final KeyValue<Windowed<KIn>, ValueAndTimestamp<VAgg>> kv = windowToEmit.next();
+                tupleForwarder.maybeForward(
+                    record.withKey(kv.key)
+                        .withValue(new Change<>(kv.value.value(), null))
+                        .withTimestamp(kv.value.timestamp())
+                        .withHeaders(null)); // Don't set header
+            }
+            emittedRecordsSensor.record(emittedCount);
+
+            lastEmitCloseTime = closeTime;
+            internalProcessorContext.addProcessorMetadataKeyValue(storeName, closeTime);
```

Comment:
After considering the logic, I am wondering if it might be simpler to not store the latest "closeTime" but the latest window-start time for which we did emit (ie, `emitRangeLowerBoundInclusive`)? It's of course just shifted by window-size and grace-period but might make the code easier to structure and read/understand?

Reply:
I thought about this and thought it might be more consistent if we shift both `lastEmitCloseTime` and `closeTime`. If we just store `lastEmitStartTime`, we only shift `closeTime` when processing. Also I feel it's easier to store `closeTime` as it's easier to see if a window closes without shifting it and see.